Refactor the transaction validity interval test

aniketd opened the pull request:
# Description

* Remove the incomplete, and now redundant `transVITimeUpperBoundIsClosed`.
* Use `toPlutusTxInfo` to comprehensively test for all protocol versions.
* The new test runs for all eras: alonzo, babbage and conway.

NOTE: This test only works for `PlutusV1` since we do not yet have common
accessors across common data types from different versions of Plutus.

Closes #5249 

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [x] Tests added or updated when needed.
- [ ] ~~`CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).~~
- [ ] ~~Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).~~
- [ ] ~~Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).~~
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

lehins commented on eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/TxInfoSpec.hs:
```diff
@@ -386,6 +388,9 @@ spec =
     txInfoSpecV2 @era SPlutusV2
+    describe "transVITime" $ do
+      it "validity interval's upper bound is closed when protocol < 9" $
+        transVITimeUpperBoundIsClosed @BabbageEra
```
What would be the point of making it era agnostic and using BabbageEra instead :smile: 
```suggestion
        transVITimeUpperBoundIsClosed @era
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -0,0 +1,55 @@
+  ) =>
+  Bool ->
+  Expectation
+transVITimeUpperBoundIsClosed' isClosed = do
```
I would suggest changing this test to ImpSpec. This way you would be able to get `ei` and `ss` from globals instead of making stuff up. Also you could adjust the interval using the slot number from the state.
Nothing says that we must submit transactions in our impspec tests.

So, instead of creating this module, it would be better to move this test to `Imp.Utxos.Valid` module.

This would mean also that you can remove any changes to cardano-ledger-babbage, because you could specify the desired behavior from the protocol version. In other words it would be one test that specifies the desired behavior for all eras

aniketd replied:
Thanks @lehins :raised_hands: Your suggestion in the meeting about using `toPlutusTxInfo` as a better alternative was awesome! Now the test is more comprehensive, as well as much better in actually testing the functionality that is used. It also now runs for all eras and tests correctly the expectation based on the protocol versions.

aniketd committed:
Refactor the transaction validity interval test.

* Remove the incomplete, and now redundant `transVITimeUpperBoundIsClosed`.
* Use `toPlutusTxInfo` to comprehensively test for all protocol versions.
* The new test runs for all eras: alonzo, babbage and conway.

NOTE: This test only works for PlutusV1 since we do not yet have common
accessors accross common data types from different versions of Plutus.

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp.hs:
```diff
@@ -36,6 +38,7 @@ spec ::
   , InjectRuleFailure "LEDGER" AlonzoUtxoPredFailure era
   , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
   , InjectRuleFailure "LEDGER" AlonzoUtxowPredFailure era
+  , EraPlutusTxInfo L.PlutusV1 era
```
Move this constraint to `AlonzoEraTest` instead, this will lead to less constraints in the `spec` for all eras

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxosSpec.hs:
```diff
@@ -40,10 +58,46 @@ spec ::
+                  ( PV1.Finite
+                      (PV1.POSIXTime expectedUpperBound)
+                  )
+                  (protVer < ProtVer (natVersion @9) 0) -- The upper bound.
```
```suggestion
                  (pvMajor protVer < natVersion @9) -- The upper bound.
```

lehins commented on eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp.hs:
```diff
@@ -44,6 +45,7 @@ spec ::
   , InjectRuleFailure "LEDGER" BabbageUtxoPredFailure era
   , InjectRuleFailure "LEDGER" BabbageUtxowPredFailure era
   , Inject (BabbageContextError era) (ContextError era)
+  , EraPlutusTxInfo L.PlutusV1 era
```
See my other comment
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp.hs:
```diff
@@ -94,6 +94,7 @@ spec ::
   , Eq (Event (EraRule "ENACT" era))
   , Typeable (Event (EraRule "ENACT" era))
   , ToExpr (Event (EraRule "BBODY" era))
+  , EraPlutusTxInfo PlutusV1 era
```
See my other comment
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp.hs:
```diff
@@ -94,6 +94,7 @@ spec ::
   , Typeable (Event (EraRule "ENACT" era))
   , ToExpr (Event (EraRule "BBODY" era))
+  , EraPlutusTxInfo PlutusV1 era
   , EraPlutusTxInfo PlutusV2 era
```
While you are at it you can move this constraint to `BabbageEraTest`
```suggestion
```

aniketd replied:
I also added `EraPlutusTxInfo PlutusV3 era` to `ConwayEraTest`, while I was at it :+1:

lehins reviewed (approved):
Very nice! Thank you!
